### ISSUE TITLE
Slow to encrypt/upload

### ISSUE THREAD

**Issue by simonjgreen** [keybase/client#1948]
Unsure if this is a bug or just a limitation. I grabbed a folder of PSDs to upload (I wanted something chunky to test with), between 1 MB and 15 MB per file. 417 MB total. It's taking an extremely long time to encrypt and upload each file, but using remarkably little CPU to do so. Took around 10 minutes total.

http://recordit.co/UW3wp1Gtkh

No bandwidth limitations that this would hit from where I am either.

It then errored halfway through saying "No data to be written"

Keybase GUI Version: 1.0.10-20160208170011+01e1c79

**Comment by strib** [comment]
Hi @simonjgreen, thanks for the report.  We definitely haven't done ANY work at all to optimize upload speeds, and in particular there's a lot we can do to optimize how much data we keep buffered depending on the network speed.  This isn't an immediate priority, but we'll get there soon (I hope).  So I'm not too surprised that uploading 400 MB takes that long.

More worrisome is the error.  It might just be a simple operation timeout due to the size of the uploads, but it's hard to know from this description. If you haven't deleted the logs, would you mind doing a `keybase log send` and posting the log ID here so we can check out the error?  Thanks!

**Comment by akalin-keybase** [comment]
I tried reproing and I got plenty of these messages:

```
2016-02-11T01:49:47.797534 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3224] c8a7 Write unblocked [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
2016-02-11T01:49:47.797611 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3188] c8a8 Blocking a write because of 16779513 dirty bytes [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
2016-02-11T01:49:47.971584 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3224] c8a9 Write unblocked [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
2016-02-11T01:49:47.971660 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3188] c8aa Blocking a write because of 16779513 dirty bytes [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
2016-02-11T01:49:48.145785 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3224] c8ab Write unblocked [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
2016-02-11T01:49:48.145863 ▶ [DEBU kbfs(FBO 025a3cf4) folder_branch_ops.go:3188] c8ac Blocking a write because of 16779513 dirty bytes [tags:FID=VXzeoxxhY8bAJqNFbsFscA]
```

It seems to go pretty fast for a while, but then hangs (no network activity) when it gets into this state.

**Comment by AppShipIt** [comment]
I'm even having trouble with a 60K PDF. So far if I retry enough times it seems to work eventually.

**Comment by grempe** [comment]
I would also chime in on this issue.  This is not really a factor of file size.  I just went through the process of updating my Jekyll blog so I could serve it from KBFS.  The blog (https://grempe.keybase.pub/www.rempe.us/about/), when generated, results in about ~125 files being uploaded each time I generate the site. The process takes about 100ms when run on a non-KBFS dir on my machine, but when writing into /keybase/public/grempe/www.rempe.us dir it takes ~3 minutes!!! 1800x longer. Every time. Most of the files are unchanged from the file it replaces.

Each file seems to take more than ~1500ms to sign and upload, and these are generally small files (no file larger than 100k I think, most 1-2kb).  Also, during this time I receive many (couple of dozen?) Notification Center notifications from OS X about signing and uploading files.  These notifications, which BTW are nearly useless and I would argue should be eliminated, do nothing but annoy since they don't even say the name of the file being signed and uploaded.

I think KBFS should:

- Only upload and re-sign files where the hash of the file changed, otherwise replacing an existing file with one that hashes to the same hash should be effectively instantaneous.

- Do the sign and sync to the server asynchronously, after the files have been copied to the FS.  This may not be possible using FUSE, I don't know.

- Provide a dropbox style UI indication to know which files are sync'ed and which have not yet been (dropbox does this with a small icon next to the filename in the finder).

- Eliminate the UI notifications from the Keybase app for signing/upload, especially if a filename is not specified.

Cheers.

**Comment by strib** [comment]
Thanks @grempe, we are already working on some of the performance aspects, including background writes.  That's going to take us a while though, because we have to be very careful with it -- it changes the semantics of the file system in ways that makes it harder to use by apps.  But we will be rolling it out as the default for people sometime this year.

As for only re-uploading if the contents change -- we already do that, although currently you have to have recently read/wrote that file from that same device (without having restarting Keybase).  Even then, though, there will still be round-trip latency and a signature because you are updating other file system metadata like the time the file was last updated.  Maybe with `rsync` you can preserve mtimes and avoid overwriting files that haven't changed?  I haven't tested it though.

Overall yes, write performance is a known issue and we are actively working on it.  We've just been focused on getting everything working correctly first.  Thanks for giving it a try!

**Comment by grempe** [comment]
Thanks @strib for the quick update.  I look forward to the enhancements.

**Comment by strib** [comment]
Version 1.0.20 has architectural changes that should greatly improve the large write experience.  If you're still seeing problems, please open a new issue.
